pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1659,3 +1675,69 @@ func (r *ReconcileClusterDeployment) deleteStaleProvisions(provs []*hivev1.Clust
+func (r *ReconcileClusterDeployment) getAllSyncSetInstances(cd *hivev1.ClusterDeployment) ([]*hivev1.SyncSetInstance, error) {
+
+	// get all syncset instances for the clusterdeployment namespace
+	ssiList := &hivev1.SyncSetInstanceList{}
+	err := r.Client.List(context.TODO(), ssiList, client.InNamespace(cd.Namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	syncSetInstances := []*hivev1.SyncSetInstance{}
+	for i, syncSetInstance := range ssiList.Items {
+		if syncSetInstance.Spec.ClusterDeployment.Name == cd.Name {
+			syncSetInstances = append(syncSetInstances, &ssiList.Items[i])
+		}
+	}
+	return syncSetInstances, nil
+}
+
+// checkFailureSyncCondition return true when syncsetinstance status condition has hivev1.ApplyFailureSyncCondition
+func checkFailureSyncCondition(syncSetInstances []*hivev1.SyncSetInstance) bool {
+	for _, syncSetInstance := range syncSetInstances {
+		for _, condition := range syncSetInstance.Status.Conditions {
+			if condition.Type == hivev1.ApplyFailureSyncCondition {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// updateSyncSetFailureCondition returns true when it sets the hivev1.SyncSetFailureCondition
+// hivev1.SyncSetFailureCondition is set if any syncsetinstance has status condition type as hivev1.ApplyFailureSyncCondition will return true
+func (r *ReconcileClusterDeployment) updateSyncSetFailureCondition(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
+
+	// get all syncset instances for this cluster deployment
+	syncSetInstances, err := r.getAllSyncSetInstances(cd)
+	if err != nil {
+		cdLog.WithError(err).Error("unable to list related sync set instances for cluster deployment")
+		return false, err
+	}
+
+	isfailedCondition := checkFailureSyncCondition(syncSetInstances)
+
+	// if failure found, add the cluster deployment status condition SyncSetFailureCondition
+	if isfailedCondition == true {
+		cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.SyncSetFailureCondition,
+			corev1.ConditionTrue, "SyncSetApplyFailure", "One of the SyncSetInstance apply has failed", controllerutils.UpdateConditionAlways)
+		err := r.Status().Update(context.TODO(), cd)
+		if err != nil {
+			cdLog.WithError(err).Error("error updating cluster deployment with SyncSetFailureCondition condition (= true)")
+			return false, err
+		}
+	} else {
+		// update the SyncSetFailureCondition if it exists on the cluster deployment
+		cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.SyncSetFailureCondition,
+			corev1.ConditionFalse, "SyncSetApplySuccess", "SyncSet apply is successful", controllerutils.UpdateConditionAlways)
+		err := r.Status().Update(context.TODO(), cd)
+		if err != nil {
+			cdLog.WithError(err).Error("error updating cluster deployment with SyncSetFailureCondition condition (= false)")
+			return false, err
+		}
+	}
+	return true, nil
```

Comment:
always returning true here for the non-error case means the caller will do an early exit of the controller. i think you want to use the results of UpdateClusterConditionWithChangeCheck()

Reply:
Right, I have refactored the code to fix it.